File: packages/expo-app-metrics/ios/Spans/SpanWriter.swift
// Copyright 2025-present 650 Industries. All rights reserved.

import ExpoModulesCore

/// The single sink for completed span rows. Every producer (the network request producer, the
/// JS spans API, future navigation spans) hands finished rows here instead of talking to the
/// database, so the persistence concerns live in one place: the actor hop, the row cap applied
/// by the insert, and failure swallowing — recording telemetry must never break a producer.
@AppMetricsActor
final class SpanWriter: Sendable {
  /// Test seam: a database injected at construction wins over the shared one. `nil` in
  /// production, where the shared database is resolved lazily on the actor — that keeps the
  /// `AppMetrics.spanWriter` static constructible from nonisolated contexts (the JS thread).
  private let databaseOverride: MetricsDatabase?

  nonisolated init() {
    self.databaseOverride = nil
  }

  /// `database` is optional so a failed database open degrades to dropped rows instead of
  /// blocking the producers.
  init(database: MetricsDatabase?) {
    self.databaseOverride = database
  }

  /// Writes one completed row. Callers already on `AppMetricsActor` (the network producer's
  /// synchronous record path) use this directly.
  func write(_ row: SpanRow) {
    guard let database = databaseOverride ?? AppMetrics.database else {
      return
    }
    do {
      try database.insert(span: row)
    } catch {
      logger.warn("[AppMetrics] Failed to persist span \"\(row.name)\": \(error.localizedDescription)")
    }
  }

  /// Fire-and-forget write from any isolation (the JS spans API calls this from the JS thread).
  nonisolated func schedule(_ row: SpanRow) {
    Task { @AppMetricsActor in
      self.write(row)
    }
  }
}

File: packages/expo-app-metrics/android/src/main/java/expo/modules/appmetrics/spans/SpanWriter.kt
// Copyright 2025-present 650 Industries. All rights reserved.

package expo.modules.appmetrics.spans

import android.util.Log
import expo.modules.appmetrics.storage.MetricsDatabase
import expo.modules.appmetrics.storage.Span
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

private const val TAG = "ExpoAppMetrics"

/**
 * The single sink for completed span rows. Every producer (the network request producer, the
 * JS spans API, future navigation spans) hands finished rows here instead of talking to the
 * database, so the persistence concerns live in one place: the coroutine hop off the caller's
 * thread, the session-FK ordering, the row cap applied by the insert, and failure swallowing —
 * recording telemetry must never break a producer. Mirrors the iOS `SpanWriter`.
 */
class SpanWriter(
  private val database: MetricsDatabase,
  private val scope: CoroutineScope,
  /**
   * Awaited before each insert so a row never races the session row it references. The main
   * session persist is idempotent, so this is a cheap no-op after the first write.
   */
  private val awaitSessionPersisted: suspend () -> Unit = {}
) {
  /**
   * Writes one completed row on `scope`, so callers (OkHttp dispatcher threads, the JS thread)
   * never block on the database.
   */
  fun write(span: Span) {
    scope.launch {
      try {
        awaitSessionPersisted()
        database.spanDao().insertCapped(span)
      } catch (e: Exception) {
        Log.w(TAG, "Failed to persist span \"${span.name}\"", e)
      }
    }
  }

  /**
   * Writes a batch produced lazily inside the write coroutine. The producer runs off the
   * caller's thread, so a large batch (the network producer's startup backfill) doesn't pay
   * its conversion cost on the thread that scheduled it. Rows fail independently — one bad
   * insert doesn't drop the rest of the batch.
   */
  fun writeAll(produce: () -> List<Span>) {
    scope.launch {
      try {
        awaitSessionPersisted()
      } catch (e: Exception) {
        Log.w(TAG, "Failed to persist a span batch", e)
        return@launch
      }
      for (span in produce()) {
        try {
          database.spanDao().insertCapped(span)
        } catch (e: Exception) {
          Log.w(TAG, "Failed to persist span \"${span.name}\"", e)
        }
      }
    }
  }
}
